### ISSUE TITLE
Progpilot doesn't detect vulnerability in WordPress plugin "Official MailerLite Sign Up Forms < 1.4.4 - Unauthenticated SQL Injection"

### ISSUE THREAD

**Issue by Mister-Stein** [designsecurity/progpilot#49]
Progpilot built from source, includes recent commit, PHP 8.2.1
Configuration:
```yml
inputs:
    dev_mode: false
    languages: 
        - php
    sources:
        keep_defaults: true
        config_files: ~
    sinks:
        keep_defaults: false
        config_files: "/home/kira/programs/progpilot_configurations/custom_configuration_new/sinks.json"
    validators:
        keep_defaults: true
        config_files: ~
    sanitizers:
        keep_defaults: true
        config_files: ~
    customrules:
        keep_defaults: true
        config_files: ~
    inclusions: ~
    resolved_includes_file: ~
    false_positives: ~
outputs:
    taintedflow: true
    include_failures_file: ~
    
options:
    analyze_includes: true
    debug_mode: true
    pretty_print: true
    max_file_analysis_duration: 1000000
    max_definitions: 300000000
    max_file_size: 50000000000
```
Contents of the `sinks.json` file are the same as in default, in the exception of these lines that I've added in the end of the file, so Progpilot could detect some sinks in methods of `$wpdb` object:
```json
{"name": "get_col", "instanceof": "wpdb", "parameters": [{"id": 1, "conditions": "QUOTES"}], "language": "php", "attack": "sql_injection", "cwe": "CWE_89"},
{"name": "get_results", "instanceof": "wpdb", "parameters": [{"id": 1, "conditions": "QUOTES"}], "language": "php", "attack": "sql_injection", "cwe": "CWE_89"},
{"name": "get_row", "instanceof": "wpdb", "parameters": [{"id": 1, "conditions": "QUOTES"}], "language": "php", "attack": "sql_injection", "cwe": "CWE_89"},
{"name": "get_var", "instanceof": "wpdb", "parameters": [{"id": 1, "conditions": "QUOTES"}], "language": "php", "attack": "sql_injection", "cwe": "CWE_89"},
{"name": "insert", "instanceof": "wpdb", "parameters": [{"id": 2, "conditions": "array_tainted"}], "language": "php", "attack": "sql_injection", "cwe": "CWE_89"},
{"name": "update", "instanceof": "wpdb", "parameters": [{"id": 2, "conditions": "array_tainted"}], "language": "php", "attack": "sql_injection", "cwe": "CWE_89"},
{"name": "query", "instanceof": "wpdb", "parameters": [{"id": 1, "conditions": "QUOTES"}], "language": "php", "attack": "sql_injection", "cwe": "CWE_89"}
```
Of course it is not the full list of possible sinks of `$wpdb`, but they are the most common.
Lets check if it works, for example, with `get_row` method in this example code:
```php
<?php
$wpdb = new wpdb("localhost", "MONA", "RED", "DEMODB");

$form = $wpdb->get_row( "SELECT * FROM " . $wpdb->base_prefix . "mailerlite_forms WHERE id = " . $_POST['form_id'] );
```
It works! And if to check in this:
```php
<?php
$wpdb = new wpdb("localhost", "MONA", "RED", "DEMODB");

function test() {
	global $wpdb;
	$form = $wpdb->get_row( "SELECT * FROM " . $wpdb->base_prefix . "mailerlite_forms WHERE id = " . $_POST['form_id'] );
}
test();
```
Cool! It obviously works, thanks to the recent commit.

Information about the vulnerability in WordPress plugin `Official MailerLite Sign Up Forms < 1.4.4 - Unauthenticated SQL Injection` may be found in these places:
WPScan: [https://wpscan.com/vulnerability/a4b58b7b-ce26-452d-bd28-37171314bd03](url)
PatchStack: [https://patchstack.com/articles/sql-injection-csrf-vulnerabilities-in-mailerlite-sign-up-forms-plugin/](url)

There's several vulnerabilities, but lets focus on simplest one, that's in `MailerLite_Admin::ajax_get_more_groups` function inside `/include/mailerlite-admin.php` file.
Vulnerable code:
```php
global $wpdb;

$form = $wpdb->get_row(
	"SELECT * FROM " . $wpdb->base_prefix
	. "mailerlite_forms WHERE id = " . $_POST['form_id']
);
```

Lets try to find this vulnerability with Progpilot. I've downloaded MailerLite 1.4.3 for the test, from official plugins directory: [https://wordpress.org/plugins/official-mailerlite-sign-up-forms/advanced/](url)

But the scan shows nothing useful. My thoughts about this:

1. Obviously, object `$wpdb` is created not inside the plugin. It is created in WordPress core, inside `require_wp_db` function in `/wp-includes/load.php` file. `$wpdb` variable is declared as global, object from `wpdb` class is created and assigned to that variable. So, that way `$wpdb` object is accessible for plugin developers.
2. Any function or method of a class in the code of any plugin is consequently, and eventually in the end, is executed by the core. Plugins are just hook up their own functions using `add_action`, so they get executed by corresponding `do_action` in the core. Of course there's `filters` too, but we're not talking about them now.
3. I was trying to include the whole WordPress core in to the analysis alongside with the plugin, but it's pretty long and generate a LOT of errors.
4. I was trying to include WordPress "stubs" in the analysis alongside with the plugin. I've generated "stubs" file using [php-stubs/generator](https://github.com/php-stubs/generator) from [johnpbloch/wordpress-core](https://github.com/johnpbloch/wordpress-core). But the problem here is that it does contain only definitions of classes of identified objects, not the actual creation of an object like that: `$wpdb = new wpdb("localhost", "MONA", "RED", "DEMODB");`. But it does contain declarations of other global variables and constants, that contain just strings, not objects.

I'm running out of ideas, so I hope I've described the situation properly.

**Comment by eric-therond** [comment]
Thank you @Mister-Stein for this detailed issue
I fixed some stuffs:
- I put your custom configuration related to wordpress [in the default configuration](https://github.com/designsecurity/progpilot/blob/master/package/src/uptodate_data/php/frameworks/wordpress/sinks.json), feel free to adjust it
- I added the [possibility to declare a variable](https://github.com/designsecurity/progpilot/blob/master/docs/CUSTOM_ANALYSIS.md#create-an-object), wherever it is found during the analysis, as an instance of an object, it means that it is now possible to declare variable `wpdb` as instance of object type `wpdb` with a custom rule also added [in the default configuration](https://github.com/designsecurity/progpilot/blob/master/package/src/uptodate_data/php/frameworks/wordpress/rules.json), and so you can analyze a standalone plugin, without wordpress core source code, and `wpdb` should be recognized as a `wpdb` object type and so the related sinks are automatically available.
- fixed some deprecations regarding php 8.2

Tell me if you have results you expect.

**Comment by Mister-Stein** [comment]
Oh my, I'm completely forgot about custom rules! You are genius! :100:  Now it detects all sources of SQLI in the plugin! Just one false positive because we haven't configured WordPress's sanitizers yet. Now I can continue the research and testing on other vulnerabilities, to help you further polish ProgPilot's capabilities. During research, I'll be opening PRs from time to time, to add more sources, sanitizers, sinks, etc. If some problem occurs, then I know how to get in touch with you. :grin: 
Seriously, ProgPilot have potential, just needs some polishing. :smiley:
